### PR TITLE
add initOnce

### PR DESCRIPTION
**Pull request by MartinNowak**
- uses double-checked locking (the correct one)

- all instances shared the same global mutex

- minimal overhead and simple API

### CODE REVIEW

File: std/concurrency.d [hunk 1]
```diff
@@ -2403,3 +2403,118 @@ version( unittest )
...
+    }
+}
+
+/**
+ * Initializes $(D_PARAM var) with the lazy $(D_PARAM init) value in a
+ * thread-safe manner.  The implementation guarantees that all threads
+ * simultaneously calling initOnce with the same $(D_PARAM var) argument block
+ * until $(D_PARAM var) is fully initialized. All side-effects of $(D_PARAM init)
+ * are globally visible afterwards.
+ *
```
Comment (MartinNowak) [inline std/concurrency.d]: Should I seriously write `var - the variable to initialize` and `init - the initializer` even though the first sentence perfectly describes what this function does? This feels a lot like `auto sum = a + b; // add a and b` comments.

Comment (WalterBright) [inline std/concurrency.d]: Yes. The main description can be more of an overview, and be more of an explanation of what the point of this function is.

Comment (MartinNowak) [inline std/concurrency.d]: Done, but I still disagree.
We don't improve our documentation by adding noise.
What's that supposed to tell me?
http://www.cplusplus.com/reference/cmath/pow/#parameters

File: std/concurrency.d [hunk 2]
```diff
@@ -2403,3 +2403,144 @@ version( unittest )
...
+
+    @property Mutex initOnceLock()
+    {
+        return lock;
+    }
+}
+
+/**
+ * Initializes $(D_PARAM var) with the lazy $(D_PARAM init) value in a
+ * thread-safe manner.
```
Comment (jacob-carlborg) [inline std/concurrency.d]: Could you please add a blank line to make the first part a summary.

File: std/concurrency.d [hunk 3]
```diff
@@ -2403,3 +2403,144 @@ version( unittest )
...
+    foreach (_; 0 .. 10)
+        spawn({ownerTid.send(MySingleton.instance.val);});
+    foreach (_; 0 .. 10)
+        assert(receiveOnly!size_t == MySingleton.instance.val);
+    assert(MySingleton.cnt == 1);
+}
+
+/**
+ * Same as above, but takes a separate mutex instead of sharing one among
+ * all initOnce instances.
```
Comment (jacob-carlborg) [inline std/concurrency.d]: Could you please add a blank line to make the first part a summary.

File: std/concurrency.d (hunk 2 above)
Comment (MartinNowak) [inline std/concurrency.d]: done

**Comment by MartinNowak** [comment]
Anything left to do?

**Comment by MartinNowak** [comment]
Ping again

**Comment by DmitryOlshansky** [comment]
LGTM

File: std/concurrency.d (hunk 3 above)
Comment (MartinNowak) [inline std/concurrency.d]: also done

**Comment by MartinNowak** [comment]
Someone merge this please?

**Comment by MartinNowak** [comment]
Auto-merge toggled on

**Pull request merged**